pkg/controllers/nsxserviceaccount/nsxserviceaccount_controller.go

```diff
@@ -0,0 +1,242 @@
+	"github.com/vmware-tanzu/nsx-operator/pkg/metrics"
+	servicecommon "github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/nsxserviceaccount"
+	nsxutil "github.com/vmware-tanzu/nsx-operator/pkg/nsx/util"
+)
+
+var (
+	log                     = logger.Log
+	ResultNormal            = common.ResultNormal
+	ResultRequeue           = common.ResultRequeue
+	ResultRequeueAfter5mins = common.ResultRequeueAfter5mins
+	MetricResType           = common.MetricResTypeNSXServiceAccount
+)
+
+// NSXServiceAccountReconciler reconciles a NSXServiceAccount object
+type NSXServiceAccountReconciler struct {
+	client.Client
+	Scheme  *apimachineryruntime.Scheme
+	Service *nsxserviceaccount.NSXServiceAccountService
+}
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
+func (r *NSXServiceAccountReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	obj := &nsxvmwarecomv1alpha1.NSXServiceAccount{}
+	log.Info("reconciling CR", "nsxserviceaccount", req.NamespacedName)
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerSyncTotal, MetricResType)
+
+	if err := r.Client.Get(ctx, req.NamespacedName, obj); err != nil {
+		log.Error(err, "unable to fetch NSXServiceAccount CR", "req", req.NamespacedName)
+		return ResultNormal, client.IgnoreNotFound(err)
+	}
+
+	// Since NSXServiceAccount service can only be activated from NSX 4.1.0 onwards,
+	// So need to check NSX version before starting NSXServiceAccount reconcile
+	if !r.Service.NSXClient.NSXCheckVersionForNSXServiceAccount() {
+		err := errors.New("NSX version check failed, NSXServiceAccount feature is not supported")
+		updateFail(r, &ctx, obj, &err)
+		// if NSX version check fails, it will be put back to reconcile queue and be reconciled after 5 minutes
+		return ResultRequeueAfter5mins, nil
+	}
+
+	if obj.ObjectMeta.DeletionTimestamp.IsZero() {
+		metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateTotal, MetricResType)
+		if !controllerutil.ContainsFinalizer(obj, servicecommon.NSXServiceAccountFinalizerName) {
+			controllerutil.AddFinalizer(obj, servicecommon.NSXServiceAccountFinalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				log.Error(err, "add finalizer", "nsxserviceaccount", req.NamespacedName)
+				updateFail(r, &ctx, obj, &err)
+				return ResultRequeue, err
+			}
+			log.V(1).Info("added finalizer on CR", "nsxserviceaccount", req.NamespacedName)
+		}
+
+		if obj.Status.Phase == nsxvmwarecomv1alpha1.NSXServiceAccountPhaseRealized {
+			return ResultNormal, nil
+		}
+		if err := r.Service.CreateOrUpdateNSXServiceAccount(ctx, obj); err != nil {
+			if errors.As(err, &nsxutil.RestrictionError{}) {
```

Comment:
I do not see RestrictionError raised from CreateOrUpdateNSXServiceAccount func.

Reply:
Thanks. This check is not needed for nsxserviceaccount. Removed.